--- /usr/local/www/netmagis/login.orig	2016-02-29 14:59:03.289990840 +0000
+++ /usr/local/www/netmagis/login	2016-02-29 14:59:09.353986525 +0000
@@ -189,7 +189,7 @@
     switch -glob $meta(Content-Type) {
 	{text/plain;*} {
 	    # CAS v1
-	    d error [mc "Invalid CAS answer"]
+	    d error [mc "Invalid CAS answer 2"]
 	}
 	{text/html;*} {
 	    # CAS v2
@@ -222,8 +222,31 @@
 
 	    welcome-user $dbfd $login $ticket
 	}
+        {application/xml;*} {
+            # CAS v2+
+            set body $state(body)
+            set query "//cas:serviceResponse/cas:authenticationSuccess/cas:user/text()"
+            dom parse $body document
+            $document documentElement root
+            set textNode [$root selectNodes $query]
+            if {$textNode eq ""} then {
+       		# validation failed, return error message
+		set query "//cas:serviceResponse/cas:authenticationFailure"
+		dom parse $body document
+		$document documentElement root
+
+		set textNode [$failureNode selectNodes text()]
+		set reason [$textNode nodeValue]
+		d error [mc "CAS login failed (%s)" $reason]
+            }
+            set login [$textNode nodeValue]
+            if {! [check-login $login]} then {
+                d error [mc "Invalid login (%s)" $login]
+            }
+            welcome-user $dbfd $login $ticket
+        }
 	{*} {
-	    d error [mc "Invalid CAS answer"]
+	    d error [mc "Invalid CAS answer (%s)" $meta(Content-Type)]
 	}
     }
     exit 0
